app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb

```diff
@@ -24,6 +24,29 @@ def host_setup_extension
 
         super
       end
+
+      def context_urls
+        find_smart_proxy
+
+        super.merge(rhsm_url: rhsm_url, pulp_content_url: pulp_content_url)
+      end
+
+      private
+
+      def find_smart_proxy
+        @smart_proxy ||= params[:url] ? SmartProxy.find_by(url: params[:url]) : SmartProxy.pulp_primary
+
+        fail Foreman::Exception, N_('Smart proxy content source not found!') unless @smart_proxy
+        fail Foreman::Exception, N_('Pulp 3 is not enabled on Smart proxy!') unless @smart_proxy.pulp3_enabled?
```

Comment:
A common pattern for this is
```suggestion
      def smart_proxy
        @smart_proxy ||= begin
          proxy = params[:url] ? SmartProxy.find_by(url: params[:url]) : SmartProxy.pulp_primary

          fail Foreman::Exception, N_('Smart proxy content source not found!') unless proxy
          fail Foreman::Exception, N_('Pulp 3 is not enabled on Smart proxy!') unless proxy.pulp3_enabled?
          
          proxy
        end
```

Then you can just call `smart_proxy.rhsm_url` without first having to call `find_smart_proxy`. Not having to rely on side effects is IMHO a good pattern.